# Lab 1 - Programming in `pandas`: Part 1

In [ ]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

Welcome to lab 1! 
In this lab, you will first download a dataset from an external source, then analyize it using basic `pandas` functions.

There are many online sources to find datasets. One of the most well known sites to find datasets is [Kaggle](https://www.kaggle.com), which hosts data science competitions and challenges. 
In fact, the dataset used in this lab will be from Kaggle. Another site where you can search for datasets is Google's own [Dataset Search Engine](https://datasetsearch.research.google.com/) 


## Cereals

Your best friend Shashank has been noticing that his normal breakfast cereal contains too much sugar per serving for his liking. 
Shashank has recently started his journey on healthy eating and excercising. 
He would like to know more about which cereal is the best in terms of nutritional value. 
Not knowing much about data anaysis, Shashank turns to you for help on finding the best cereal. 

A dataset about cereal and their nutritional value can be found here: https://www.kaggle.com/crawford/80-cereals

You can get an overall understanding of what this dataset records by reading the description of the data. 

###  Question 1 - Getting your Data
After reading the dataset description, complete the following steps:

1. Download the dataset as a `csv` from Kaggle. You will need to make a Kaggle account to download the dataset.
2. Unzip the data into your current directory, if the file downloads as a `zip`
3. Rename the `csv` file to be `cereal.csv` and upload it to your current directory

```
BEGIN QUESTION
name: q1
```

Now, we can read in the `csv` as a dataframe using `pandas`.
Assign the filepath to `dataset_filepath` as a string and load the cereal dataframe into `cereal`.

Since you should have uploaded the file to the same folder where this notebook is located, `dataset_filepath` should just be the name of the `csv`.

In [ ]:
""" # BEGIN PROMPT
# File name 
dataset_filepath = ...

# Read in the file using pandas
cereal = ...
cereal.head()
# END PROMPT
"""
# BEGIN SOLUTION NO PROMPT
## SOLUTION ##
# File name 
dataset_filepath = 'cereal.csv'

# Read in the file using pandas
cereal = pd.read_csv(dataset_filepath)
cereal.head()
# END SOLUTION

name mfr type  calories  protein  fat  sodium  fiber  \
0                  100% Bran   N    C        70        4    1     130   10.0   
1          100% Natural Bran   Q    C       120        3    5      15    2.0   
2                   All-Bran   K    C        70        4    1     260    9.0   
3  All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   
4             Almond Delight   R    C       110        2    2     200    1.0   

   carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0    5.0       6     280        25      3     1.0  0.33  68.402973  
1    8.0       8     135         0      3     1.0  1.00  33.983679  
2    7.0       5     320        25      3     1.0  0.33  59.425505  
3    8.0       0     330        25      3     1.0  0.50  93.704912  
4   14.0       8      -1        25      3     1.0  0.75  34.384843

In [ ]:
## Test ##
assert len(cereal.columns) == 16

### Question 2 - Understanding your Data

**Question 2.1:** How extensive will your research be using this dataset? 
To answer this question you need to find out how many different types of cereal we have in this dataset. 
Assign `num_types` to the total number of cereals recorded.


```
BEGIN QUESTION
name: q2_1
```

In [ ]:
""" # BEGIN PROMPT

num_types = ...
num_types

# END PROMPT """

# BEGIN SOLUTION NO PROMPT

## SOLUTION ##
num_types = cereal.shape[0]
num_types

# END SOLUTION

77

In [ ]:
## SOLUTION ##
num_types = cereal.shape[0]
num_types

77

In [ ]:
## Test ##

assert num_types == 77

**Question 2.2:** How many columns does our dataset have? Assign `num_columns` to the number of columns in our dataset.

```
BEGIN QUESTION
name: q2_2

In [ ]:
""" # BEGIN PROMPT

num_columns = ...
num_columns

# END PROMPT """

# BEGIN SOLUTION NO PROMPT

## SOLUTION ##
num_columns = cereal.shape[1]
num_columns

# END SOLUTION

16

In [ ]:
## SOLUTION ##
num_columns = cereal.shape[1]
num_columns

16

In [ ]:
## Test ##

assert num_columns == 16

**Question 2.3:** To get a general understanding of the data, you would like to look at the first 20 cereals listed in this dataset. Assign 
`first_twenty` to a dataframe that contains the first 20 rows of the cereal dataset. 

```
BEGIN QUESTION
name: q2_3

In [ ]:
""" # BEGIN PROMPT
first_twenty = ...
# END PROMPT """

# BEGIN SOLUTION NO PROMPT 

## SOLUTION ##
first_twenty = cereal.head(20)
first_twenty

# END SOLUTION

name mfr type  calories  protein  fat  sodium  fiber  \
0                   100% Bran   N    C        70        4    1     130   10.0   
1           100% Natural Bran   Q    C       120        3    5      15    2.0   
2                    All-Bran   K    C        70        4    1     260    9.0   
3   All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   
4              Almond Delight   R    C       110        2    2     200    1.0   
5     Apple Cinnamon Cheerios   G    C       110        2    2     180    1.5   
6                 Apple Jacks   K    C       110        2    0     125    1.0   
7                     Basic 4   G    C       130        3    2     210    2.0   
8                   Bran Chex   R    C        90        2    1     200    4.0   
9                 Bran Flakes   P    C        90        3    0     210    5.0   
10               Cap'n'Crunch   Q    C       120        1    2     220    0.0   
11                   Cheerios   G    C       110        6    2     290    2.0   
12      Cinnamon Toast Crunch   G    C       120        1    3     210    0.0   
13                   Clusters   G    C       110        3    2     140    2.0   
14                Cocoa Puffs   G    C       110        1    1     180    0.0   
15                  Corn Chex   R    C       110        2    0     280    0.0   
16                Corn Flakes   K    C       100        2    0     290    1.0   
17                  Corn Pops   K    C       110        1    0      90    1.0   
18              Count Chocula   G    C       110        1    1     180    0.0   
19         Cracklin' Oat Bran   K    C       110        3    3     140    4.0   

    carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0     5.0       6     280        25      3    1.00  0.33  68.402973  
1     8.0       8     135         0      3    1.00  1.00  33.983679  
2     7.0       5     320        25      3    1.00  0.33  59.425505  
3     8.0       0     330        25      3    1.00  0.50  93.704912  
4    14.0       8      -1        25      3    1.00  0.75  34.384843  
5    10.5      10      70        25      1    1.00  0.75  29.509541  
6    11.0      14      30        25      2    1.00  1.00  33.174094  
7    18.0       8     100        25      3    1.33  0.75  37.038562  
8    15.0       6     125        25      1    1.00  0.67  49.120253  
9    13.0       5     190        25      3    1.00  0.67  53.313813  
10   12.0      12      35        25      2    1.00  0.75  18.042851  
11   17.0       1     105        25      1    1.00  1.25  50.764999  
12   13.0       9      45        25      2    1.00  0.75  19.823573  
13   13.0       7     105        25      3    1.00  0.50  40.400208  
14   12.0      13      55        25      2    1.00  1.00  22.736446  
15   22.0       3      25        25      1    1.00  1.00  41.445019  
16   21.0       2      35        25      1    1.00  1.00  45.863324  
17   13.0      12      20        25      2    1.00  1.00  35.782791  
18   12.0      13      65        25      2    1.00  1.00  22.396513  
19   10.0       7     160        25      3    1.00  0.50  40.448772

In [ ]:
## SOLUTION ##
first_twenty = cereal.head(20)
first_twenty

name mfr type  calories  protein  fat  sodium  fiber  \
0                   100% Bran   N    C        70        4    1     130   10.0   
1           100% Natural Bran   Q    C       120        3    5      15    2.0   
2                    All-Bran   K    C        70        4    1     260    9.0   
3   All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   
4              Almond Delight   R    C       110        2    2     200    1.0   
5     Apple Cinnamon Cheerios   G    C       110        2    2     180    1.5   
6                 Apple Jacks   K    C       110        2    0     125    1.0   
7                     Basic 4   G    C       130        3    2     210    2.0   
8                   Bran Chex   R    C        90        2    1     200    4.0   
9                 Bran Flakes   P    C        90        3    0     210    5.0   
10               Cap'n'Crunch   Q    C       120        1    2     220    0.0   
11                   Cheerios   G    C       110        6    2     290    2.0   
12      Cinnamon Toast Crunch   G    C       120        1    3     210    0.0   
13                   Clusters   G    C       110        3    2     140    2.0   
14                Cocoa Puffs   G    C       110        1    1     180    0.0   
15                  Corn Chex   R    C       110        2    0     280    0.0   
16                Corn Flakes   K    C       100        2    0     290    1.0   
17                  Corn Pops   K    C       110        1    0      90    1.0   
18              Count Chocula   G    C       110        1    1     180    0.0   
19         Cracklin' Oat Bran   K    C       110        3    3     140    4.0   

    carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0     5.0       6     280        25      3    1.00  0.33  68.402973  
1     8.0       8     135         0      3    1.00  1.00  33.983679  
2     7.0       5     320        25      3    1.00  0.33  59.425505  
3     8.0       0     330        25      3    1.00  0.50  93.704912  
4    14.0       8      -1        25      3    1.00  0.75  34.384843  
5    10.5      10      70        25      1    1.00  0.75  29.509541  
6    11.0      14      30        25      2    1.00  1.00  33.174094  
7    18.0       8     100        25      3    1.33  0.75  37.038562  
8    15.0       6     125        25      1    1.00  0.67  49.120253  
9    13.0       5     190        25      3    1.00  0.67  53.313813  
10   12.0      12      35        25      2    1.00  0.75  18.042851  
11   17.0       1     105        25      1    1.00  1.25  50.764999  
12   13.0       9      45        25      2    1.00  0.75  19.823573  
13   13.0       7     105        25      3    1.00  0.50  40.400208  
14   12.0      13      55        25      2    1.00  1.00  22.736446  
15   22.0       3      25        25      1    1.00  1.00  41.445019  
16   21.0       2      35        25      1    1.00  1.00  45.863324  
17   13.0      12      20        25      2    1.00  1.00  35.782791  
18   12.0      13      65        25      2    1.00  1.00  22.396513  
19   10.0       7     160        25      3    1.00  0.50  40.448772

In [ ]:
## Test ##

assert len(first_twenty.columns) == 16
assert len(first_twenty.index) == 20
assert first_twenty.iloc[10,10] == 35
assert first_twenty.iloc[1,4] == 3

### Question 3 - Indexing your Data

**Question 3.1:** Shashank's current favorite cereal is `Fruity Pebbles`. He would like to know the amount of calories and sugar per serving for this cereal. 

Assign `pebbles` to a dataframe containing only the amount of calories and sugars per serving of `Fruity Pebbles`.

```
BEGIN QUESTION
name: q3_1
```

In [ ]:
""" # BEGIN PROMPT

pebbles = ...
pebbles

# END PROMPT """

# BEGIN SOLUTION NO PROMPT

## SOLUTION ##
pebbles = cereal[cereal["name"] == 'Fruity Pebbles'].loc[:, ['calories', 'sugars']]
pebbles

# END SOLUTION

calories  sugars
29       110      12

In [ ]:
## SOLUTION ##
pebbles = cereal[cereal["name"] == 'Fruity Pebbles'].loc[:, ['calories', 'sugars']]
pebbles

calories  sugars
29       110      12

In [ ]:
## Test ##

assert len(pebbles.columns) == 2
assert len(pebbles.index) == 1
assert pebbles.iloc[0, 1] == 12
assert pebbles.iloc[0, 0] == 110

**Question 3.2:** What cereal(s) has the most amount of protein (and consequently is the best for Shashank's gains)? Assign `most_protein` to the name of the cereal with the most protein. If there are multiple cereals with the most protein, you can assign `most_protein` to any of the cereals.

_Hint:_ Remeber, if you are doing `Series[5]`, you are extracting the value with index 5, and not the 5th element. 

```
BEGIN QUESTION
name: q3_2

In [ ]:
""" ##BEGIN PROMPT
most_protein = ...
most_protein

## END PROMPT """

# BEGIN SOLUTION NO PROMPT

## Solution ##
most_protein = cereal.sort_values(by = "protein", ascending = False)["name"].values[0]
most_protein

# END SOLUTION

'Cheerios'

In [ ]:
## Solution ##
most_protein = cereal.sort_values(by = "protein", ascending = False)["name"].values[0]
most_protein

'Cheerios'

In [ ]:
## Test ##

assert most_protein == 'Cheerios'

**Question 3.3:** Shashank says that he is also interested in maximizing fiber intake for breakfast.
He asks you to find all the cereals with greater than or equal to 4 grams of protein and more than 3 grams of fiber. 
Assign `qualified_cereals` to all the cereals that have greater than or equal to 4 grams of protein and more than 3 grams of fiber. 

``` 
BEGIN QUESTION
name: q3_3
```

In [ ]:
""" ## BEGIN PROMPT

qualified_cereals = ...
qualified_cereals

## END PROMPT """

# BEGIN SOLUTION NO PROMPT
## SOLUTION ##
qualified_cereals = cereal.loc[(cereal['protein'] >= 4) & (cereal['fiber'] > 3)]
qualified_cereals
# END SOLUTION

name mfr type  calories  protein  fat  sodium  fiber  \
0                  100% Bran   N    C        70        4    1     130   10.0   
2                   All-Bran   K    C        70        4    1     260    9.0   
3  All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   

   carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0    5.0       6     280        25      3     1.0  0.33  68.402973  
2    7.0       5     320        25      3     1.0  0.33  59.425505  
3    8.0       0     330        25      3     1.0  0.50  93.704912

In [ ]:
## SOLUTION ##
qualified_cereals = cereal.loc[(cereal['protein'] >= 4) & (cereal['fiber'] > 3)]
qualified_cereals

name mfr type  calories  protein  fat  sodium  fiber  \
0                  100% Bran   N    C        70        4    1     130   10.0   
2                   All-Bran   K    C        70        4    1     260    9.0   
3  All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   

   carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0    5.0       6     280        25      3     1.0  0.33  68.402973  
2    7.0       5     320        25      3     1.0  0.33  59.425505  
3    8.0       0     330        25      3     1.0  0.50  93.704912

In [ ]:
qualified_cereals['protein'].to_numpy()


array([4, 4, 4])

In [ ]:
## Test ##
assert len(qualified_cereals.index) == 3
assert np.count_nonzero(qualified_cereals['protein'].to_numpy() >= 4) == 3
assert np.count_nonzero(qualified_cereals['fiber'].to_numpy() > 3) == 3

**Question 3.4:** Suppose we are only interested in the name, protein, fiber, calories, vitamins, and sugar values of all the cereals. 
Use `.loc` to return a dataframe with all the cereals containing only our columns of interest in the order presented above. Assign this dataframe to the variable `cereal_loc`.

```
BEGIN QUESTION
name: q3_4
```

In [ ]:
""" ## BEGIN PROMPT
cereal_loc = ...
cereal_loc
## END PROMPT """

# BEGIN SOLUTION NO PROMPT
# SOLUTION
cereal_loc = cereal.loc[:, ['name', 'protein', 'fiber', 'calories', 'vitamins', 'sugars']]
cereal_loc

# END SOLUTION

name  protein  fiber  calories  vitamins  sugars
0                   100% Bran        4   10.0        70        25       6
1           100% Natural Bran        3    2.0       120         0       8
2                    All-Bran        4    9.0        70        25       5
3   All-Bran with Extra Fiber        4   14.0        50        25       0
4              Almond Delight        2    1.0       110        25       8
..                        ...      ...    ...       ...       ...     ...
72                    Triples        2    0.0       110        25       3
73                       Trix        1    0.0       110        25      12
74                 Wheat Chex        3    3.0       100        25       3
75                   Wheaties        3    3.0       100        25       3
76        Wheaties Honey Gold        2    1.0       110        25       8

[77 rows x 6 columns]

In [ ]:
cereal_loc = cereal.loc[:, ['name', 'protein', 'fiber', 'calories', 'vitamins', 'sugars']]
cereal_loc

name  protein  fiber  calories  vitamins  sugars
0                   100% Bran        4   10.0        70        25       6
1           100% Natural Bran        3    2.0       120         0       8
2                    All-Bran        4    9.0        70        25       5
3   All-Bran with Extra Fiber        4   14.0        50        25       0
4              Almond Delight        2    1.0       110        25       8
..                        ...      ...    ...       ...       ...     ...
72                    Triples        2    0.0       110        25       3
73                       Trix        1    0.0       110        25      12
74                 Wheat Chex        3    3.0       100        25       3
75                   Wheaties        3    3.0       100        25       3
76        Wheaties Honey Gold        2    1.0       110        25       8

[77 rows x 6 columns]

In [ ]:
## Test ##

assert len(cereal_loc.index) == 77
assert len(cereal_loc.columns) == 6
assert cereal_loc.columns[0] == 'name'
assert cereal_loc.columns[1] == 'protein'
assert cereal_loc.columns[2] == 'fiber'
assert cereal_loc.columns[3] == 'calories'
assert cereal_loc.columns[4] == 'vitamins'
assert cereal_loc.columns[5] == 'sugars'

**Question 3.5:** After seeing `cereal_loc`, Shashank tells you that he is not interested in the vitamin content of different cereals. Use the  `.drop` funftion to drop all the columns that we do not want. 
Assign this dataframe to the variable `cereal_drop`.

*Hint:* Make sure you are dropping labels along the column axis.

```
BEGIN QUESTION
name: q3_5

In [ ]:
""" ## BEGIN PROMPT
cereal_drop = ...
cereal_drop

## END PROMPT """

# BEGIN SOLUTION NO PROMPT
## SOLUTION ##
cereal_drop = cereal_loc.drop(labels = ['vitamins'], axis = 1)
cereal_drop
# END SOLUTION

name  protein  fiber  calories  sugars
0                   100% Bran        4   10.0        70       6
1           100% Natural Bran        3    2.0       120       8
2                    All-Bran        4    9.0        70       5
3   All-Bran with Extra Fiber        4   14.0        50       0
4              Almond Delight        2    1.0       110       8
..                        ...      ...    ...       ...     ...
72                    Triples        2    0.0       110       3
73                       Trix        1    0.0       110      12
74                 Wheat Chex        3    3.0       100       3
75                   Wheaties        3    3.0       100       3
76        Wheaties Honey Gold        2    1.0       110       8

[77 rows x 5 columns]

In [ ]:
## SOLUTION ##
cereal_drop = cereal_loc.drop(labels = ['vitamins'], axis = 1)
cereal_drop

name  protein  fiber  calories  sugars
0                   100% Bran        4   10.0        70       6
1           100% Natural Bran        3    2.0       120       8
2                    All-Bran        4    9.0        70       5
3   All-Bran with Extra Fiber        4   14.0        50       0
4              Almond Delight        2    1.0       110       8
..                        ...      ...    ...       ...     ...
72                    Triples        2    0.0       110       3
73                       Trix        1    0.0       110      12
74                 Wheat Chex        3    3.0       100       3
75                   Wheaties        3    3.0       100       3
76        Wheaties Honey Gold        2    1.0       110       8

[77 rows x 5 columns]

In [ ]:
## Test ##

assert 'vitamins' not in cereal_drop.columns
assert len(cereal_drop.columns) == 5
assert len(cereal_drop.index) == 77

**Question 3.6:** Shashank brings back from Trader Joe's some Honey O's. He tells you the following nutritional information:
- Name: Honey O's
- Protein: 2 g
- Fiber; 2 g
- Calories: 110 kcal
- Sugars: 9 g

Add this new observation into your table `cereal_drop` from 3.5.

*Hint:* replace the `...` after `.loc` with the index that you want the new row to be assigned to, and fill in the right hand side of the assignment as a list containing all nessecary nutritional information. The list should have 1 value for each column of the dataframe.

```
BEGIN QUESTION
name: q3_6

In [ ]:
""" ## BEGIN PROMPT
cereal_drop.loc[...] = ...
cereal_drop

## END PROMPT """

# BEGIN SOLUTION NO PROMPT
## SOLUTION ##
cereal_drop.loc[77] = ["Honey O's", 2, 2, 110, 9]
cereal_drop
# END PROMPT

name  protein  fiber  calories  sugars
0                   100% Bran        4   10.0        70       6
1           100% Natural Bran        3    2.0       120       8
2                    All-Bran        4    9.0        70       5
3   All-Bran with Extra Fiber        4   14.0        50       0
4              Almond Delight        2    1.0       110       8
..                        ...      ...    ...       ...     ...
73                       Trix        1    0.0       110      12
74                 Wheat Chex        3    3.0       100       3
75                   Wheaties        3    3.0       100       3
76        Wheaties Honey Gold        2    1.0       110       8
77                  Honey O's        2    2.0       110       9

[78 rows x 5 columns]

In [ ]:
## SOLUTION ##
cereal_drop.loc[77] = ["Honey O's", 2, 2, 110, 9]
cereal_drop

name  protein  fiber  calories  sugars
0                   100% Bran        4   10.0        70       6
1           100% Natural Bran        3    2.0       120       8
2                    All-Bran        4    9.0        70       5
3   All-Bran with Extra Fiber        4   14.0        50       0
4              Almond Delight        2    1.0       110       8
..                        ...      ...    ...       ...     ...
73                       Trix        1    0.0       110      12
74                 Wheat Chex        3    3.0       100       3
75                   Wheaties        3    3.0       100       3
76        Wheaties Honey Gold        2    1.0       110       8
77                  Honey O's        2    2.0       110       9

[78 rows x 5 columns]

In [ ]:
cereal_drop.iloc[77, 0]

"Honey O's"

In [ ]:
## Test ##

assert len(cereal_drop.index) == 78
assert cereal_drop.iloc[77, 0] == "Honey O's"
assert cereal_drop.iloc[77, 1] == 2
assert cereal_drop.iloc[77, 2] == 2
assert cereal_drop.iloc[77, 3] == 110
assert cereal_drop.iloc[77, 4] == 9


### Question 4

You realize that the `cups` column has different cups per serving, so that the nutritional information is not standardized across cereals.
Since you normally eat 1 cup of cereal for breakfast, it would be better to see the nutritional values for each cereal per cup. 

Let's transform the data so that the dataframe reflects the nutritional values of each cereal per cup and store it in a new table called `serving`. `serving` should have all the columns of `cereal_drop`, but the `'calories', 'protein', 'fiber', 'sugars'` columns should reflect serving per 1 cup. 

*Hint 1:* For example, the serving size for cheerios is 1.25 and the caloric content is 110. We can divide calories by cups (110 / 1.25) to get calories per cup. Similarly, All-Bran with Extra Fiber has 50 calories for 0.5 (or 1/2) cups of serving. Could we divide calories by cups here as well?

As a challenge, you can try using `df.div`. You can find the documentation [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.divide.html). Don't forget the `axis` argument.

```
BEGIN QUESTION
name: q4
```

In [ ]:
""" ## BEGIN PROMPT
serving = ...
serving

## END PROMPT """

# BEGIN SOLUTION NO PROMPT
## SOLUTION ##
serving = cereal[['name', 'protein', 'fiber', 'calories', 'sugars', 'cups']]
serving['protein'] = serving['protein'] / serving['cups']
serving['fiber'] = serving['fiber'] / serving['cups']
serving['calories'] = serving['calories'] / serving['cups']
serving['sugars'] = serving['sugars'] / serving['cups']
serving = serving.drop('cups', axis = 1)
serving
# END SOLUTION

name    protein      fiber    calories     sugars
0                   100% Bran  12.121212  30.303030  212.121212  18.181818
1           100% Natural Bran   3.000000   2.000000  120.000000   8.000000
2                    All-Bran  12.121212  27.272727  212.121212  15.151515
3   All-Bran with Extra Fiber   8.000000  28.000000  100.000000   0.000000
4              Almond Delight   2.666667   1.333333  146.666667  10.666667
..                        ...        ...        ...         ...        ...
72                    Triples   2.666667   0.000000  146.666667   4.000000
73                       Trix   1.000000   0.000000  110.000000  12.000000
74                 Wheat Chex   4.477612   4.477612  149.253731   4.477612
75                   Wheaties   3.000000   3.000000  100.000000   3.000000
76        Wheaties Honey Gold   2.666667   1.333333  146.666667  10.666667

[77 rows x 5 columns]

In [ ]:
## SOLUTION ##
serving = cereal[['name', 'protein', 'fiber', 'calories', 'sugars', 'cups']]
serving['protein'] = serving['protein'] / serving['cups']
serving['fiber'] = serving['fiber'] / serving['cups']
serving['calories'] = serving['calories'] / serving['cups']
serving['sugars'] = serving['sugars'] / serving['cups']
serving = serving.drop('cups', axis = 1)
serving

name    protein      fiber    calories     sugars
0                   100% Bran  12.121212  30.303030  212.121212  18.181818
1           100% Natural Bran   3.000000   2.000000  120.000000   8.000000
2                    All-Bran  12.121212  27.272727  212.121212  15.151515
3   All-Bran with Extra Fiber   8.000000  28.000000  100.000000   0.000000
4              Almond Delight   2.666667   1.333333  146.666667  10.666667
..                        ...        ...        ...         ...        ...
72                    Triples   2.666667   0.000000  146.666667   4.000000
73                       Trix   1.000000   0.000000  110.000000  12.000000
74                 Wheat Chex   4.477612   4.477612  149.253731   4.477612
75                   Wheaties   3.000000   3.000000  100.000000   3.000000
76        Wheaties Honey Gold   2.666667   1.333333  146.666667  10.666667

[77 rows x 5 columns]

In [ ]:
## ALTERNATE SOLUTION ##
serving = cereal[['name', 'protein', 'fiber', 'calories', 'sugars', 'cups']]
serving[['protein', 'fiber', 'calories', 'sugars', 'cups']] = serving[['protein', 'fiber', 'calories', 'sugars', 'cups']].div(serving['cups'], axis=0)
serving = serving.drop('cups', axis = 1)
serving

name    protein      fiber    calories     sugars
0                   100% Bran  12.121212  30.303030  212.121212  18.181818
1           100% Natural Bran   3.000000   2.000000  120.000000   8.000000
2                    All-Bran  12.121212  27.272727  212.121212  15.151515
3   All-Bran with Extra Fiber   8.000000  28.000000  100.000000   0.000000
4              Almond Delight   2.666667   1.333333  146.666667  10.666667
..                        ...        ...        ...         ...        ...
72                    Triples   2.666667   0.000000  146.666667   4.000000
73                       Trix   1.000000   0.000000  110.000000  12.000000
74                 Wheat Chex   4.477612   4.477612  149.253731   4.477612
75                   Wheaties   3.000000   3.000000  100.000000   3.000000
76        Wheaties Honey Gold   2.666667   1.333333  146.666667  10.666667

[77 rows x 5 columns]

In [ ]:
## Test ##

assert len(serving.index) == 77
assert len(serving.columns) == 5
assert np.isclose(12.121212121, serving.iloc[0,1])
assert np.isclose(27.27272727, serving.iloc[2, 2])


## Question 5

Shashank has tasted all of the different cereals and given them a rating on a scale of 1 (least) to 5(most). The information is stored in the file `ratings.csv`. Your task is to load the file and merge with with the cereal dataset such that each cereal 

```
BEGIN QUESTION
name: q5
```

In [ ]:
ratings = ...

In [ ]:
## SOLUTION ##
ratings = pd.read_csv("ratings.csv")
cereal  = cereal.merge(ratings, on = "name")
cereal

name mfr type  calories  protein  fat  sodium  fiber  \
0                   100% Bran   N    C        70        4    1     130   10.0   
1           100% Natural Bran   Q    C       120        3    5      15    2.0   
2                    All-Bran   K    C        70        4    1     260    9.0   
3   All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   
4              Almond Delight   R    C       110        2    2     200    1.0   
..                        ...  ..  ...       ...      ...  ...     ...    ...   
72                    Triples   G    C       110        2    1     250    0.0   
73                       Trix   G    C       110        1    1     140    0.0   
74                 Wheat Chex   R    C       100        3    1     230    3.0   
75                   Wheaties   G    C       100        3    1     200    3.0   
76        Wheaties Honey Gold   G    C       110        2    1     200    1.0   

    carbo  sugars  potass  vitamins  shelf  weight  cups     rating  \
0     5.0       6     280        25      3     1.0  0.33  68.402973   
1     8.0       8     135         0      3     1.0  1.00  33.983679   
2     7.0       5     320        25      3     1.0  0.33  59.425505   
3     8.0       0     330        25      3     1.0  0.50  93.704912   
4    14.0       8      -1        25      3     1.0  0.75  34.384843   
..    ...     ...     ...       ...    ...     ...   ...        ...   
72   21.0       3      60        25      3     1.0  0.75  39.106174   
73   13.0      12      25        25      2     1.0  1.00  27.753301   
74   17.0       3     115        25      1     1.0  0.67  49.787445   
75   17.0       3     110        25      1     1.0  1.00  51.592193   
76   16.0       8      60        25      1     1.0  0.75  36.187559   

    shashank_ratings  
0                  2  
1                  3  
2                  2  
3                  4  
4                  3  
..               ...  
72                 2  
73                 4  
74                 2  
75                 3  
76                 1  

[77 rows x 17 columns]

In [ ]:
## Test ##

assert len(cereal.columns) == 17
assert 'shashank_ratings' in cereal.columns
assert 'rating' in cereal.columns

## GroupBy

Shashank browses his favorite nutrition blog and sees several articles about the dangers of high sugar consumption. 
Unfortunately, he also has a big sweet tooth and is worried that he only likes eating cereals that are particularly sweet.

Use `.groupby` to return a dataframe with a row for each rating and the average sugar content for the cereals that were given that rating. 
Make sure that your final dataframe has only one column, `mean_sugars`.

Assign the dataframe to `mean_sugar_per_rating`.

```
BEGIN QUESTION
name: q6

In [0]:
mean_sugar_per_rating = ...
mean_sugar_per_rating

In [ ]:
## SOLUTION ##

mean_sugar_per_rating = cereal[['shashank_ratings', 'sugars']].groupby('shashank_ratings').mean()
mean_sugar_per_rating = mean_sugar_per_rating.rename(columns={'sugars' : 'mean_sugars'})
mean_sugar_per_rating

mean_sugars
shashank_ratings             
1                    7.000000
2                    6.600000
3                    7.411765
4                    6.916667
5                    6.600000

In [ ]:
## Test ##

assert len(mean_sugar_per_rating.columns) == 1
assert 'mean_sugars' in mean_sugar_per_rating.columns
assert np.isclose(7, mean_sugar_per_rating.iloc[0,0])
assert np.isclose(7.4117647, mean_sugar_per_rating.iloc[2,0])

## Pivot

After giving each cereal his personal rating, Shashank realizes that he wants to also get the cereal with the most
protein as well. Shashank is trying to balance his personal taste rating with the amount of protein in each cereal.
In the following cell, create a pivot table that compares `shashank_ratings` versus `protein` to visually aid us
in balancing Shashank's personal taste and the amount of protein in each cereal. Assign this pivot table to `pivot_df`.

```
BEGIN QUESTION
name: q7
```

In [0]:
pivot_df = ...
pivot_df

In [ ]:
## SOLUTION ##

pivot_df = pd.pivot_table(cereal, values='name', index = 'protein', columns = 'shashank_ratings', aggfunc=list)
pivot_df

shashank_ratings                                                  1  \
protein                                                               
1                 [Cap'n'Crunch, Cinnamon Toast Crunch, Count Ch...   
2                 [Apple Jacks, Corn Chex, Froot Loops, Just Rig...   
3                   [Post Nat. Raisin Bran, Shredded Wheat 'n'Bran]   
4                                                     [Life, Maypo]   
5                                                  [Quaker Oatmeal]   
6                                                        [Cheerios]   

shashank_ratings                                                  2  \
protein                                                               
1                                      [Golden Grahams, Honey-comb]   
2                      [Shredded Wheat, Total Corn Flakes, Triples]   
3                 [Basic 4, Frosted Mini-Wheats, Fruit & Fibre D...   
4                                             [100% Bran, All-Bran]   
5                                                               NaN   
6                                                               NaN   

shashank_ratings                                                  3  \
protein                                                               
1                                                     [Cocoa Puffs]   
2                 [Almond Delight, Multi-Grain Cheerios, Rice Kr...   
3                 [100% Natural Bran, Bran Flakes, Cream of Whea...   
4                 [Muesli Raisins; Peaches; & Pecans, Quaker Oat...   
5                                                               NaN   
6                                                               NaN   

shashank_ratings                                                  4  \
protein                                                               
1                               [Fruity Pebbles, Puffed Rice, Trix]   
2                                          [Bran Chex, Double Chex]   
3                 [Clusters, Cracklin' Oat Bran, Grape Nuts Flak...   
4                 [All-Bran with Extra Fiber, Muesli Raisins; Da...   
5                                                               NaN   
6                                                               NaN   

shashank_ratings                                                  5  
protein                                                              
1                                     [Corn Pops, Honey Graham Ohs]  
2                 [Apple Cinnamon Cheerios, Corn Flakes, Crispix...  
3                 [Grape-Nuts, Just Right Fruit & Nut, Product 1...  
4                                                               NaN  
5                                                               NaN  
6                                                       [Special K]

In [ ]:
## Test ##

assert "Cap'n'Crunch" in pivot_df.iloc[0,0]
assert 'Golden Grahams' in pivot_df.iloc[0,1]
assert 'Life' in pivot_df.iloc[3, 0]
assert 'Special K' in pivot_df.iloc[5,4]

* merge build our own data, export it, and ask students to import and join it (fabricated rating by Shashank np.random.rand_int )  **Shashank**
* groupby our own rating, counting the number of instances on each rating **David**
* pivot (less emphasis) on rating and protein to find the best rated cereal with the most protein.  **Ervin**

Recommend it to Shashank **Grace**